### PR TITLE
Refactor Relations to not be per-EventTimelineSet

### LINKED ISSUE #2399: EventTimelineSet constructor and docs are contradictory
**Describe the bug**
The docblock of `EventTimelineSet`s constructor states that the room should be nullable, the parameters however state that it is not. Also the usage within the class is not clear how it should be used. (`getPendingEvents` has a guard clause, while `aggregateRelations` does not)

This leads to errors if you use the Timelineset without providing a room.

**To Reproduce**
1. Create a new EventTimelineSet with room set to null
2. Try to add an Event with a relation (in our case `m.replace`)

**Expected behavior**
The class should work correctly without a room or the docblock should be changed.

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser: chrome, firefox

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/matrix-org/matrix-js-sdk/issues/2399
Fixes https://github.com/vector-im/element-web/issues/22298

All the "breaking" changes are on code which was marked as experimental so does not need a major version bump.

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Refactor Relations to not be per-EventTimelineSet ([\#2412](https://github.com/matrix-org/matrix-js-sdk/pull/2412)). Fixes #2399 and vector-im/element-web#22298.

### CODE REVIEW

**Comment by t3chguy** [comment]
Blocking until after RC

File: src/models/relations-container.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    // this.relations[relatesToEventId][relationType][relationEventType]
+    private relations: {
+        [relatesToEventId: string]: {
+            [relationType: RelationType | string]: {
+                [eventType: EventType | string]: Relations;
+            };
+        };
+    } = {};
+
+    constructor(private readonly client: MatrixClient, private readonly room?: Room) {
```
Comment (turt2live) [inline src/models/relations-container.ts:35]: `client` is nullable thanks to how event-timeline-set works on the constructor (and possibly other places)

File: src/models/relations-container.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Get a collection of relations to a given event in this timeline set.
```
Comment (turt2live) [inline src/models/relations-container.ts:39]: There's a reason why the spec decided to abandon all hope of documenting what a "relation event", "related event", and "relatable event" are: it gets quite complicated. 


Having reviewed this PR, I think it's quite important that we match the spec on this as it's nearly impossible to understand what event is being talked about with the current overlap in vocabulary. The spec uses "parent" and "child" to denote the relationship between the events, described here: https://github.com/matrix-org/matrix-spec/pull/1062

File: src/models/relations-container.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+     */
+    public getRelationsForEvent(
+        eventId: string,
+        relationType: RelationType | string,
+        eventType: EventType | string,
+    ): Relations | undefined {
+        return this.relations[eventId]?.[relationType]?.[eventType];
+    }
+
+    public getAllRelationsEventForEvent(eventId: string): MatrixEvent[] {
```
Comment (turt2live) [inline src/models/relations-container.ts:63]:
```suggestion
    public getAllRelationsEventsForEvent(eventId: string): MatrixEvent[] {
```

.. or are they related events? :P

**Review by turt2live — CHANGES_REQUESTED** [review]
generally looks sane - just a few nitpicks and a whole naming scheme change 😇

File: src/models/relations-container.ts (hunk 1 above)
Comment (t3chguy) [reply src/models/relations-container.ts:35]: It is required, the jsdoc for EventTimelineSet says client or room must be provided (or both, but that isn't stated)

**Review by turt2live — APPROVED** [review]

**Pull request merged**